arangod/Aql/SharedQueryState.h

```diff
@@ -101,10 +101,27 @@ class SharedQueryState final
   /// execute a task in parallel if capacity is there
   template<typename F>
   bool asyncExecuteAndWakeup(F&& cb) {
+    // The atomic _numTasks counts the number of ongoing tasks. When it
+    // drops to 0, we need to wake up a thread which is waiting for this
+    // on the condition variable _cv. We must not miss this event, or
+    // ellse we might have a thread which is waiting forever.
+    // The waiting thread uses a predicate to check if _numTasks is 0,
+    // and only goes to sleep when it is not. This happens under the
+    // mutex _mutex and releasing the mutex and going to sleep is an
+    // atomic operation. Thus, to not miss the event that _numTasks
+    // is reduced to zero, we must, whenever we decrement it, do this
+    // under the mutex, and then, after releasing the mutex, notify the
+    // condition variable _cv! Then either the decrement or the going to
+    // sleep happens first (serialized by the mutex). If the decrement
+    // happens first, the waiting thread is not even going to sleep, if
+    // the going to sleep happens first, then we will wakt it up.
     unsigned num = _numTasks.fetch_add(1);
     if (num + 1 > _maxTasks) {
-      _numTasks.fetch_sub(1);  // revert
       std::forward<F>(cb)(false);
+      std::unique_lock<std::mutex> guard(_mutex);
+      _numTasks.fetch_sub(1);  // revert
+      guard.unlock();
+      _cv.notify_all();
       return false;
```

Comment:
I may have misunderstood this part but I read it as: If this task did not make progress, we are signaling others to try to make progress even though this one did not.
My concern here is why would we be notifying others to wake up if this task could not proceed. Since we just decremented the `_numTasks` that we increased previously.

Reply:
Completely agreed, I will try to investigate this. Thanks!